taxcalc/tests/test_pufcsv.py

```diff
@@ -80,19 +80,21 @@ def test_agg(tests_path, puf_path):  # pylint: disable=redefined-outer-name
         raise ValueError(msg)
     # create aggregate diagnostic table using sub sample of records
     fullsample = pd.read_csv(puf_path)
-    rn_seed = 80  # to ensure two-percent sub-sample is always the same
-    subsample = fullsample.sample(frac=0.02,  # pylint: disable=no-member
+    rn_seed = 80  # to ensure sub-sample is always the same
+    subfrac = 0.02  # sub-sample fraction
+    subsample = fullsample.sample(frac=subfrac,  # pylint: disable=no-member
                                   random_state=rn_seed)
     rec_subsample = Records(data=subsample)
     calc_subsample = Calculator(policy=Policy(), records=rec_subsample)
     adt_subsample = multiyear_diagnostic_table(calc_subsample, num_years=nyrs)
     # compare combined tax liability from full and sub samples for each year
     taxes_subsample = adt_subsample.loc["Combined Liability ($b)"]
-    reltol = 0.01  # maximum allowed relative difference in tax liability
+    reltol = 0.04  # maximum allowed relative difference in tax liability
```

Comment:
@MattHJensen, I'm not sure.  I hoped that there was something going on in the test function.  I looked at using the sample weights for the sub-sample selection and running the test with different seeds.  However, this hasn't changed the results.  With different seeds, the same 2% sample rate, the same 0.04 tolerance and no weights, this test fails about 20% of the time.  If I use the weights, this test fails every time with a relative difference ranging from about 0.07 to 0.15.  I haven't been able to figure out why using weights causes the the difference to increase.  

I still think that there could be something going on with the random seed.  Perhaps, we had a very small relative difference last time because we got lucky and drew a favorable sample.

Also, why do we expect the results from the full sample and 2% of the sample to be within 1% of each other?

I altered the function some to get these results:
```
@pytest.mark.requires_pufcsv
def test_agg(tests_path, puf_path):  # pylint: disable=redefined-outer-name
    """
    Test Tax-Calculator aggregate taxes with no policy reform using
    the full-sample puf.csv and a two-percent sub-sample of puf.csv
    """
    use_weights = False
    rn_seed = 0
    msgs = []
    n_tests = 20
    for i in range(0, n_tests):
        msg = ""
        rn_seed = np.random.randint(0,10000)  # to ensure sub-sample is always the same
        # pylint: disable=too-many-locals,too-many-statements
        nyrs = 10
        # create a Policy object (clp) containing current-law policy parameters
        clp = Policy()
        # create a Records object (rec) containing all puf.csv input records
        rec = Records(data=puf_path)
        # create a Calculator object using clp policy and puf records
        calc = Calculator(policy=clp, records=rec)
        calc_start_year = calc.current_year
        # create aggregate diagnostic table (adt) as a Pandas DataFrame object
        adt = multiyear_diagnostic_table(calc, nyrs)
        taxes_fullsample = adt.loc["Combined Liability ($b)"]
        # convert adt results to a string with a trailing EOL character
        adtstr = adt.to_string() + '\n'
        # generate differences between actual and expected results
        actual = adtstr.splitlines(True)
        aggres_path = os.path.join(tests_path, 'pufcsv_agg_expect.txt')
        with open(aggres_path, 'r') as expected_file:
            txt = expected_file.read()
        expected_results = txt.rstrip('\n\t ') + '\n'  # cleanup end of file txt
        expected = expected_results.splitlines(True)
        diff = difflib.unified_diff(expected, actual,
                                    fromfile='expected', tofile='actual', n=0)
        # convert diff generator into a list of lines:
        diff_lines = list()
        for line in diff:
            diff_lines.append(line)
        # test failure if there are any diff_lines
        if len(diff_lines) > 0:
            new_filename = '{}{}'.format(aggres_path[:-10], 'actual.txt')
            with open(new_filename, 'w') as new_file:
                new_file.write(adtstr)
            msg = 'PUFCSV AGG RESULTS DIFFER FOR FULL-SAMPLE\n'
            msg += '-------------------------------------------------\n'
            msg += '--- NEW RESULTS IN pufcsv_agg_actual.txt FILE ---\n'
            msg += '--- if new OK, copy pufcsv_agg_actual.txt to  ---\n'
            msg += '---                 pufcsv_agg_expect.txt     ---\n'
            msg += '---            and rerun test.                ---\n'
            msg += '-------------------------------------------------\n'
            raise ValueError(msg)
        # create aggregate diagnostic table using sub sample of records
        fullsample = pd.read_csv(puf_path)
        subfrac = 0.02  # sub-sample fraction
        if use_weights:  # normalize weights for sample
            fullsample['weights'] = fullsample.s006/fullsample.s006.sum()
            weights = 'weights'
        else: # set weights to None and pandas uses uniform dist to select samp
            weights = None
        subsample = fullsample.sample(frac=subfrac,  # pylint: disable=no-member
                                      random_state=rn_seed,
                                      weights=weights)
        weight_prop = subsample.s006.sum()/fullsample.s006.sum()
        rec_subsample = Records(data=subsample)
        calc_subsample = Calculator(policy=Policy(), records=rec_subsample)
        adt_subsample = multiyear_diagnostic_table(calc_subsample, num_years=nyrs)
        # compare combined tax liability from full and sub samples for each year
        taxes_subsample = adt_subsample.loc["Combined Liability ($b)"]
        reltol = 0.04  # maximum allowed relative difference in tax liability
        if not np.allclose(taxes_subsample, taxes_fullsample,
                           atol=0.0, rtol=reltol):
            msg = 'PUFCSV AGG RESULTS DIFFER IN SUB-SAMPLE AND FULL-SAMPLE\n'
            msg += 'WHEN subfrac = {:.3f} and reltol = {:.4f} '
            msg += 'and weight prop = {:.4f} and seed = {:.0f}\n'
            msg = msg.format(subfrac, reltol, weight_prop, rn_seed)
            it_sub = np.nditer(taxes_subsample, flags=['f_index'])
            it_all = np.nditer(taxes_fullsample, flags=['f_index'])
            while not it_sub.finished:
                cyr = it_sub.index + calc_start_year
                tax_sub = float(it_sub[0])
                tax_all = float(it_all[0])
                reldiff = abs(tax_sub - tax_all) / abs(tax_all)
                if reldiff > reltol:
                    msgstr = ' year,sub,full,reldif= {}\t{:.2f}\t{:.2f}\t{:.4f}\n'
                    msg += msgstr.format(cyr, tax_sub, tax_all, reldiff)
                it_sub.iternext()
                it_all.iternext()

        if len(msg) > 0:
            msgs.append(msg)
    if len(msgs) > 0:
        acc = (len(msgs)/float(n_tests))*100
        acc_msg = "Failure rate: {:.2f}%".format(acc)
        raise ValueError('\n'.join([acc_msg] + msgs))
```